collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java

```diff
@@ -0,0 +1,352 @@
+
+            MultipartBody.Builder multipartBuilder = new MultipartBody.Builder()
+                    .setType(MultipartBody.FORM)
+                    .addPart(MultipartBody.Part.createFormData("xml_submission_file", submissionFile.getName(), requestBody));
+
+            Timber.i("added xml_submission_file: %s", submissionFile.getName());
+            byteCount += submissionFile.length();
+
+            for (; fileIndex < fileList.size(); fileIndex++) {
+                File file = fileList.get(fileIndex);
+
+                String mime = mimeTypeMap.getMimeTypeFromExtension(FileUtils.getFileExtension(file.getName()));
+
+                RequestBody fileRequestBody = RequestBody.create(MediaType.parse(mime), file);
+
+                multipartBuilder.addPart(MultipartBody.Part.create(fileRequestBody));
+
+                byteCount += file.length();
+                Timber.i("added file of type '%s' %s", mime, file.getName());
+
+                // we've added at least one attachment to the request...
+                if (fileIndex + 1 < fileList.size()) {
+                    if ((fileIndex - lastFileIndex + 1 > 100) || (byteCount + fileList.get(fileIndex + 1).length()
+                            > 10000000L)) {
+                        // the next file would exceed the 10MB threshold...
+                        Timber.i("Extremely long post is being split into multiple posts");
+                        multipartBuilder.addPart(MultipartBody.Part.createFormData("*isIncomplete*", "yes"));
+                        ++fileIndex; // advance over the last attachment added...
+                        break;
+                    }
+                }
+            }
+
+            MultipartBody multipartBody = multipartBuilder.build();
+
+            try {
+                OkHttpClient client = createOkHttpClient(credentials);
+                Request request = postRequest(uri, multipartBody);
+                Response response = client.newCall(request).execute();
+
+                messageParser = new ResponseMessageParser(
+                        response.toString(),
+                        response.code(),
+                        response.message());
+
+                discardEntityBytes(response);
+
+                if (response.code() != HttpURLConnection.HTTP_CREATED && response.code() != HttpURLConnection.HTTP_ACCEPTED) {
+                    return messageParser;
+                }
+            } catch (IOException e) {
+                Timber.e(e);
+                String msg = e.getMessage();
+                if (msg == null) {
+                    msg = e.toString();
+                }
+
+                throw new IOException(msg);
+            }
+        }
+
+        return messageParser;
+    }
+
+    private OkHttpClient createOkHttpClient(@Nullable HttpCredentialsInterface credentials) {
+        OkHttpClient.Builder builder = new OkHttpClient.Builder();
```

Comment:
@xsteelej it would be good if the `okHttpClient` was provided via Dagger so it would be easier to share it across multiple connections and also handle it's scope if the need arises.

Reply:
Hi @jd-alexander, I'm not sure if there would be an advantage in Dependency Injecting OkHttpClient - The intention of the Refactor was to isolate the underlying Http component in one place. Using Dagger would mean we'd have to start importing it else where etc.

Reply:
Agreed that the approach here is ok for now. I also propose some simplifications to how it's set in https://github.com/opendatakit/collect/compare/master...lognaturel:collect-620.